File: apps/api/src/app/simulation/action_capabilities.py
"""Actions with an existing, verified Timecue write representation."""

from collections.abc import Mapping, Sequence

TIMECUE_DATE_ACTIONS = frozenset(
    {"date_shift", "schedule_shift", "weather_shift", "material_shift"}
)


def executable_actions(actions: object) -> bool:
    """Require the whole simulated bundle; never silently discard manual steps."""
    return (
        isinstance(actions, Sequence)
        and not isinstance(actions, (str, bytes))
        and bool(actions)
        and all(
            isinstance(action, Mapping)
            and action.get("type") in TIMECUE_DATE_ACTIONS
            and bool(action.get("taskId"))
            and bool(action.get("startsAt"))
            and bool(action.get("endsAt"))
            for action in actions
        )
    )

File: apps/frontend/src/lib/action-capabilities.ts
import type { Strategy } from "./types";

const dateActions = new Set([
  "date_shift",
  "schedule_shift",
  "weather_shift",
  "material_shift",
]);

export function isTimecueExecutable(strategy: Strategy): boolean {
  return (
    Boolean(strategy.actions?.length) &&
    strategy.actions!.every(
      (action) =>
        typeof action.type === "string" &&
        dateActions.has(action.type) &&
        Boolean(action.taskId && action.startsAt && action.endsAt),
    )
  );
}
